File: JavaScript/Ex2.js
// PALETA DE CORES
const ui = {
    reset: "\x1b[0m",
    bold: "\x1b[1m",
    deepPurple: "\x1b[38;5;54m",  
    violet: "\x1b[38;5;93m",      
    white: "\x1b[97m"
};

function header(titulo) {
    console.log("\n" + ui.deepPurple + ui.bold + "┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓");
    console.log(`┃  ${titulo.padEnd(30)} ┃`);
    console.log("┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛" + ui.reset);
}

function divider() {
    console.log(ui.violet + "   ────────────────────────────────" + ui.reset);
}

// PRODUTOS ABSTRATOS
class CartaoCredito {
    pagar(valor) {
        throw new Error("pagar() deve ser implementado");
    }
}

class Boleto {
    pagar(valor) {
        throw new Error("pagar() deve ser implementado");
    }
}

// PRODUTO CONCRETO - BANCO A
class CartaoBancoA extends CartaoCredito {
    pagar(valor) {
        return `💳 Banco A processou R$ ${valor.toFixed(2)} no Cartão`;
    }
}

class BoletoBancoA extends Boleto {
    pagar(valor) {
        return `🧾 Banco A emitiu boleto de R$ ${valor.toFixed(2)}`;
    }
}

// PRODUTO CONCRETO - BANCO B
class CartaoBancoB extends CartaoCredito {
    pagar(valor) {
        return `💳 Banco B autorizou R$ ${valor.toFixed(2)} no Cartão`;
    }
}

class BoletoBancoB extends Boleto {
    pagar(valor) {
        return `🧾 Banco B gerou boleto de R$ ${valor.toFixed(2)}`;
    }
}

// FÁBRICA ABSTRATA
class FabricaBanco {
    criarCartao() {
        throw new Error("criarCartao() deve ser implementado");
    }

    criarBoleto() {
        throw new Error("criarBoleto() deve ser implementado");
    }
}

// FÁBRICAS CONCRETAS
class FabricaBancoA extends FabricaBanco {
    criarCartao() {
        return new CartaoBancoA();
    }

    criarBoleto() {
        return new BoletoBancoA();
    }
}

class FabricaBancoB extends FabricaBanco {
    criarCartao() {
        return new CartaoBancoB();
    }

    criarBoleto() {
        return new BoletoBancoB();
    }
}

// CLIENTE
function finalizarCompra(fabrica, valor) {
    const cartao = fabrica.criarCartao();
    const boleto = fabrica.criarBoleto();

    console.log(ui.white + "   ➤ " + cartao.pagar(valor) + ui.reset);
    console.log(ui.white + "   ➤ " + boleto.pagar(valor) + ui.reset);
    divider();
}

// USO
header("🛒 LOJA ONLINE | BANCO A");
finalizarCompra(new FabricaBancoA(), 420.75);

header("🛒 LOJA ONLINE | BANCO B");
finalizarCompra(new FabricaBancoB(), 189.30);

File: Criacionais/AbstractMethod/python/Ex2.py
# PALETA DE CORES
class UI:
    reset = "\033[0m"
    bold = "\033[1m"
    deepPurple = "\033[38;5;54m"
    violet = "\033[38;5;93m"
    white = "\033[97m"

ui = UI()

def header(titulo):
    print("\n" + ui.deepPurple + ui.bold + "┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓")
    print(f"┃  {titulo.ljust(30)} ┃")
    print("┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛" + ui.reset)

def divider():
    print(ui.violet + " ────────────────────────────────" + ui.reset)

# PRODUTOS ABSTRATOS
class CartaoCredito:
    def pagar(self, valor):
        raise NotImplementedError("pagar() deve ser implementado")

class Boleto:
    def pagar(self, valor):
        raise NotImplementedError("pagar() deve ser implementado")

# PRODUTO CONCRETO - BANCO A
class CartaoBancoA(CartaoCredito):
    def pagar(self, valor):
        return f"💳 Banco A processou R$ {valor:.2f} no Cartão"

class BoletoBancoA(Boleto):
    def pagar(self, valor):
        return f"🧾 Banco A emitiu boleto de R$ {valor:.2f}"

# PRODUTO CONCRETO - BANCO B
class CartaoBancoB(CartaoCredito):
    def pagar(self, valor):
        return f"💳 Banco B autorizou R$ {valor:.2f} no Cartão"

class BoletoBancoB(Boleto):
    def pagar(self, valor):
        return f"🧾 Banco B gerou boleto de R$ {valor:.2f}"

# FÁBRICA ABSTRATA
class FabricaBanco:
    def criar_cartao(self):
        raise NotImplementedError("criar_cartao() deve ser implementado")

    def criar_boleto(self):
        raise NotImplementedError("criar_boleto() deve ser implementado")

# FÁBRICAS CONCRETAS
class FabricaBancoA(FabricaBanco):
    def criar_cartao(self):
        return CartaoBancoA()

    def criar_boleto(self):
        return BoletoBancoA()

class FabricaBancoB(FabricaBanco):
    def criar_cartao(self):
        return CartaoBancoB()

    def criar_boleto(self):
        return BoletoBancoB()

# CLIENTE
def finalizar_compra(fabrica, valor):
    cartao = fabrica.criar_cartao()
    boleto = fabrica.criar_boleto()

    print(ui.white + "   ➤ " + cartao.pagar(valor) + ui.reset)
    print(ui.white + "   ➤ " + boleto.pagar(valor) + ui.reset)
    divider()

# USO
header("🛒 LOJA ONLINE | BANCO A")
finalizar_compra(FabricaBancoA(), 420.75)

header("🛒 LOJA ONLINE | BANCO B")
finalizar_compra(FabricaBancoB(), 189.30)